File: src/uncomment/extract.py
"""Tree-sitter based comment extraction.

Produces a SourceFile with logical comments (adjacent line comments merged),
each classified by kind (line/block/doc) and attachment (file header,
preceding code, trailing code, or floating).
"""

from __future__ import annotations

import re
from pathlib import Path

from tree_sitter_language_pack import get_parser

from .directives import is_cgo_preamble, is_directive_text
from .languages import COMMENT_NODE_TYPES, LangSpec, spec_for_path
from .model import Attachment, Comment, FunctionInfo, Kind, SourceFile

_NAME_NODE_TYPES = frozenset(
    {"identifier", "field_identifier", "type_identifier", "property_identifier", "destructor_name", "operator_name"}
)

# a header comment directly above one of these lines documents the file, not the line
_IMPORT_LINE_RE = re.compile(
    r"^\s*(#\s*(include|pragma|ifndef|define)\b|import\b|package\b|using\b|use\b|extern crate\b|mod\b|module\b|['\"]use strict)"
)


def _function_name(node) -> str:
    name_node = node.child_by_field_name("name")
    if name_node is not None:
        return name_node.text.decode("utf-8", "replace")
    declarator = node.child_by_field_name("declarator")
    if declarator is not None:
        stack = [declarator]
        while stack:
            n = stack.pop(0)
            if n.type in _NAME_NODE_TYPES:
                return n.text.decode("utf-8", "replace")
            stack = list(n.children) + stack
    return "<anonymous>"


def strip_markers(raw: str) -> str:
    """Remove comment markers, keeping the human text."""
    out: list[str] = []
    for line in raw.splitlines():
        s = line.strip()
        for prefix in ("/*!", "/**", "/*", "//!", "///", "//"):
            if s.startswith(prefix):
                s = s[len(prefix):]
                break
        else:
            if s.startswith("*") and not s.startswith("*/"):
                s = s[1:]
        if s.endswith("*/"):
            s = s[:-2]
        out.append(s.strip())
    # drop leading/trailing empty lines produced by /* and */ on their own lines
    while out and not out[0]:
        out.pop(0)
    while out and not out[-1]:
        out.pop()
    return "\n".join(out)


class _RawComment:
    __slots__ = ("text", "start_row", "start_col", "end_row", "end_col", "func_name", "in_function")

    def __init__(self, node, func_name: str, in_function: bool):
        self.text = node.text.decode("utf-8", "replace").rstrip("\r\n")
        self.start_row = node.start_point[0]
        self.start_col = node.start_point[1]
        self.end_row = node.end_point[0]
        self.end_col = node.end_point[1]
        # some grammars (rust doc comments) include the trailing newline
        if self.end_col == 0 and self.end_row > self.start_row:
            self.end_row -= 1
            self.end_col = 1 << 30
        self.func_name = func_name
        self.in_function = in_function


def extract_source(path: str, source: str, spec: LangSpec) -> SourceFile:
    parser = get_parser(spec.grammar)
    data = source.encode("utf-8")
    tree = parser.parse(data)
    lines = source.splitlines()

    raw_comments: list[_RawComment] = []
    functions: list[FunctionInfo] = []
    # first named non-comment node starting at each row (outermost wins)
    row_first_node: dict[int, str] = {}

    stack: list[tuple[object, str, bool]] = [(tree.root_node, "", False)]
    while stack:
        node, func_name, in_func = stack.pop()
        if node.type in COMMENT_NODE_TYPES:
            raw_comments.append(_RawComment(node, func_name, in_func))
            continue
        if node.is_named and node.type != "translation_unit":
            row_first_node.setdefault(node.start_point[0], node.type)
        child_func, child_in = func_name, in_func
        if node.type in spec.function_nodes:
            name = _function_name(node)
            body = node.child_by_field_name("body")
            if body is not None:
                functions.append(
                    FunctionInfo(
                        path=path,
                        name=name,
                        start_line=node.start_point[0] + 1,
                        end_line=node.end_point[0] + 1,
                        body_line_count=body.end_point[0] - body.start_point[0] + 1,
                    )
                )
            child_func, child_in = name, True
        # push children in reverse so we visit them in source order
        for child in reversed(node.children):
            stack.append((child, child_func, child_in))

    raw_comments.sort(key=lambda c: (c.start_row, c.start_col))

    # rows occupied by comments, and per-row comment span masks
    comment_rows: set[int] = set()
    masks: dict[int, list[tuple[int, int]]] = {}
    for rc in raw_comments:
        for row in range(rc.start_row, rc.end_row + 1):
            comment_rows.add(row)
            line_len = len(lines[row]) if row < len(lines) else 0
            a = rc.start_col if row == rc.start_row else 0
            b = rc.end_col if row == rc.end_row else line_len
            masks.setdefault(row, []).append((a, b))

    code_rows: set[int] = set()
    for row, line in enumerate(lines):
        chars = list(line)
        for a, b in masks.get(row, ()):
            for i in range(a, min(b, len(chars))):
                chars[i] = " "
        if "".join(chars).strip():
            code_rows.add(row)

    first_code_row = min(code_rows) if code_rows else None

    def code_before(rc: _RawComment) -> str:
        line = lines[rc.start_row] if rc.start_row < len(lines) else ""
        return line[: rc.start_col].strip()

    def code_after(rc: _RawComment) -> str:
        line = lines[rc.end_row] if rc.end_row < len(lines) else ""
        return line[rc.end_col:].strip()

    # ---- group adjacent line comments into logical comments ----
    groups: list[list[_RawComment]] = []
    for rc in raw_comments:
        is_line = rc.text.startswith("//")
        trailing = bool(code_before(rc))
        if (
            groups
            and is_line
            and not trailing
            and groups[-1][-1].text.startswith("//")
            and not code_before(groups[-1][-1])
            and rc.start_row == groups[-1][-1].end_row + 1
            and rc.start_col == groups[-1][-1].start_col
            and _doc_class(rc.text, spec) == _doc_class(groups[-1][-1].text, spec)
            # a directive line never merges with prose, so the prose part
            # stays judged and the directive stays protected
            and _directive_line(rc.text, spec) == _directive_line(groups[-1][-1].text, spec)
        ):
            groups[-1].append(rc)
        else:
            groups.append([rc])

    comments: list[Comment] = []
    for group in groups:
        first, last = group[0], group[-1]
        raw_text = "\n".join(g.text for g in group)
        trailing_code = code_before(first)
        after_code = code_after(last)

        next_row_code = (last.end_row + 1) in code_rows
        next_line = lines[last.end_row + 1].strip() if last.end_row + 1 < len(lines) else ""
        before_first_code = first_code_row is None or first.start_row < first_code_row

        if trailing_code:
            attachment = Attachment.TRAILING
            attached = trailing_code
        elif after_code:
            attachment = Attachment.PRECEDING
            attached = after_code
        elif raw_text.startswith(("//!", "/*!")):
            # inner/module doc: documents the file, not the next item
            attachment = Attachment.FILE_HEADER
            attached = ""
        elif before_first_code and (not next_row_code or _IMPORT_LINE_RE.match(next_line)):
            attachment = Attachment.FILE_HEADER
            attached = ""
        elif next_row_code:
            attachment = Attachment.PRECEDING
            attached = next_line
        else:
            attachment = Attachment.FLOATING
            attached = ""

        kind = _classify_kind(raw_text, spec)
        if (
            kind is not Kind.DOC
            and attachment in (Attachment.PRECEDING, Attachment.FILE_HEADER)
            and not first.in_function
            and spec.doc_by_convention_nodes
            and row_first_node.get(last.end_row + 1) in spec.doc_by_convention_nodes
        ):
            # e.g. Go: a comment directly above a declaration or the package
            # clause is a doc comment, even without special markers
            kind = Kind.DOC

        content = strip_markers(raw_text)
        content_head = content.splitlines()[0] if content else ""
        comments.append(
            Comment(
                path=path,
                lang=spec.name,
                kind=kind,
                attachment=attachment,
                text=raw_text,
                content=content,
                start_line=first.start_row + 1,
                end_line=last.end_row + 1,
                col=first.start_col,
                attached_code=attached,
                in_function=first.in_function,
                function_name=first.func_name,
                is_directive=(
                    is_directive_text(content_head, spec.name, kind)
                    or is_cgo_preamble(spec.name, attached)
                ),
            )
        )

    return SourceFile(
        path=path,
        lang=spec.name,
        lines=lines,
        comments=comments,
        functions=functions,
        code_line_count=len(code_rows),
        comment_line_count=len(comment_rows),
    )


def _directive_line(text: str, spec: LangSpec) -> bool:
    return is_directive_text(strip_markers(text.splitlines()[0]), spec.name, Kind.LINE)


def _doc_class(text: str, spec: LangSpec) -> str:
    """Grouping class of a line comment: '//!' and '///' must not merge."""
    for prefix in spec.doc_line_prefixes:
        if text.startswith(prefix):
            return prefix
    return ""


def _classify_kind(text: str, spec: LangSpec) -> Kind:
    if spec.doc_block_prefixes and text.startswith(spec.doc_block_prefixes):
        # bare "/**/" or "/***/" separators are not docs
        if not text.startswith("/**/"):
            return Kind.DOC
    if spec.doc_line_prefixes and text.startswith(spec.doc_line_prefixes):
        return Kind.DOC
    if text.startswith("/*"):
        return Kind.BLOCK
    return Kind.LINE


def extract_file(path: str | Path) -> SourceFile | None:
    """Extract comments from a file; returns None for unsupported extensions."""
    p = Path(path)
    spec = spec_for_path(str(p))
    if spec is None:
        return None
    source = p.read_text(encoding="utf-8", errors="replace")
    return extract_source(str(p), source, spec)

File: rust/src/extract.rs
//! Tree-sitter comment extraction — first slice of the port.
//!
//! Parity with src/unwaffle/extract.py so far: raw comment collection,
//! adjacent line-comment merging (same column, consecutive rows, no code
//! before), comment-byte masking for code_lines/code_line_count, and
//! attachment classification. Not yet ported: docstrings, directive
//! classification, function context, doc-by-convention, the
//! directive-line merge splits.

use crate::languages::{is_comment_node, spec_for_path, LangSpec};
use crate::model::{Attachment, Comment, Kind, SourceFile};
use tree_sitter::{Node, Parser};

struct RawComment {
    text: String,
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
}

/// Split on \n only: tree-sitter rows do the same, while a \f or U+2028
/// aware split would desync row arithmetic.
fn byte_lines(data: &[u8]) -> Vec<&[u8]> {
    data.split(|&b| b == b'\n').collect()
}

pub fn strip_markers(raw: &str, line_marker: &str) -> String {
    let mut out: Vec<String> = Vec::new();
    for line in raw.split('\n') {
        let mut s = line.trim();
        if line_marker == "#" {
            // one marker only: '#### section' keeps its banner characters
            s = s.strip_prefix('#').unwrap_or(s);
        } else {
            let mut stripped = false;
            for prefix in ["/*!", "/**", "/*", "//!", "///", "//"] {
                if let Some(rest) = s.strip_prefix(prefix) {
                    s = rest;
                    stripped = true;
                    break;
                }
            }
            if !stripped && s.starts_with('*') && !s.starts_with("*/") {
                s = &s[1..];
            }
            s = s.strip_suffix("*/").unwrap_or(s);
        }
        out.push(s.trim().to_string());
    }
    while out.first().is_some_and(|l| l.is_empty()) {
        out.remove(0);
    }
    while out.last().is_some_and(|l| l.is_empty()) {
        out.pop();
    }
    out.join("\n")
}

fn collect_comments(root: Node, data: &[u8]) -> Vec<RawComment> {
    let mut out = Vec::new();
    let mut stack = vec![root];
    while let Some(node) = stack.pop() {
        if is_comment_node(node.kind()) {
            let text = node
                .utf8_text(data)
                .unwrap_or_default()
                .trim_end_matches(['\r', '\n'])
                .to_string();
            let (mut end_row, mut end_col) = (node.end_position().row, node.end_position().column);
            // some grammars include the trailing newline in the node
            if end_col == 0 && end_row > node.start_position().row {
                end_row -= 1;
                end_col = usize::MAX;
            }
            out.push(RawComment {
                text,
                start_row: node.start_position().row,
                start_col: node.start_position().column,
                end_row,
                end_col,
            });
            continue;
        }
        for i in (0..node.child_count() as u32).rev() {
            if let Some(child) = node.child(i) {
                stack.push(child);
            }
        }
    }
    out.sort_by_key(|c| (c.start_row, c.start_col));
    out
}

pub fn extract_source(path: &str, source: &str, spec: &'static LangSpec) -> SourceFile {
    let (_, language) = spec_for_path(path).expect("caller checked the language");
    let mut parser = Parser::new();
    parser
        .set_language(&language)
        .expect("grammar/runtime version mismatch is a build error, not a runtime state");
    let source = source.strip_prefix('\u{feff}').unwrap_or(source);
    let data = source.as_bytes();
    let tree = parser.parse(data, None).expect("tree-sitter returns a tree for any input");

    let raw = collect_comments(tree.root_node(), data);
    let blines = byte_lines(data);
    let lines: Vec<String> = blines
        .iter()
        .map(|bl| {
            String::from_utf8_lossy(bl)
                .trim_end_matches('\r')
                .to_string()
        })
        .collect();

    // per-row comment span masks, byte offsets
    let mut masks: Vec<Vec<(usize, usize)>> = vec![Vec::new(); blines.len()];
    let mut comment_rows = vec![false; blines.len()];
    for rc in &raw {
        for row in rc.start_row..=rc.end_row.min(blines.len().saturating_sub(1)) {
            comment_rows[row] = true;
            let a = if row == rc.start_row { rc.start_col } else { 0 };
            let b = if row == rc.end_row { rc.end_col } else { blines[row].len() };
            masks[row].push((a, b));
        }
    }

    let mut code_lines = Vec::with_capacity(blines.len());
    let mut code_rows = vec![false; blines.len()];
    for (row, bl) in blines.iter().enumerate() {
        let mut buf = bl.to_vec();
        for &(a, b) in &masks[row] {
            for byte in buf.iter_mut().take(b.min(bl.len())).skip(a) {
                *byte = b' ';
            }
        }
        let text = String::from_utf8_lossy(&buf).trim_end_matches('\r').to_string();
        code_rows[row] = !text.trim().is_empty();
        code_lines.push(text);
    }
    let first_code_row = code_rows.iter().position(|&c| c);

    // merge runs of line comments: consecutive rows, same column, own-line
    let is_line = |rc: &RawComment| rc.text.starts_with(spec.line_marker);
    let own_line = |rc: &RawComment| {
        blines[rc.start_row][..rc.start_col.min(blines[rc.start_row].len())]
            .iter()
            .all(|b| b.is_ascii_whitespace())
    };
    let mut groups: Vec<Vec<&RawComment>> = Vec::new();
    for rc in &raw {
        let merges = is_line(rc)
            && own_line(rc)
            && groups.last().is_some_and(|g| {
                let prev = g.last().unwrap();
                is_line(prev)
                    && own_line(prev)
                    && rc.start_row == prev.end_row + 1
                    && rc.start_col == prev.start_col
            });
        if merges {
            groups.last_mut().unwrap().push(rc);
        } else {
            groups.push(vec![rc]);
        }
    }

    let mut comments = Vec::new();
    for group in &groups {
        let first = group[0];
        let last = group[group.len() - 1];
        let text: String = group.iter().map(|rc| rc.text.as_str()).collect::<Vec<_>>().join("\n");
        let head = first.text.trim_start();
        let kind = if spec.doc_block_prefixes.iter().any(|p| head.starts_with(p))
            || spec.doc_line_prefixes.iter().any(|p| head.starts_with(p))
        {
            Kind::Doc
        } else if is_line(first) {
            Kind::Line
        } else {
            Kind::Block
        };

        let end_row = last.end_row;
        let attachment = if !own_line(first) {
            Attachment::Trailing
        } else if first_code_row.is_none_or(|fc| first.start_row < fc) {
            Attachment::FileHeader
        } else if end_row + 1 < code_rows.len() && code_rows[end_row + 1] {
            Attachment::Preceding
        } else {
            Attachment::Floating
        };
        let attached_code = match attachment {
            Attachment::Trailing => code_lines[first.start_row].trim().to_string(),
            Attachment::Preceding => code_lines[end_row + 1].trim().to_string(),
            _ => String::new(),
        };

        comments.push(Comment {
            path: path.to_string(),
            lang: spec.name,
            kind,
            attachment,
            content: strip_markers(&text, spec.line_marker),
            text,
            start_line: first.start_row + 1,
            end_line: end_row + 1,
            col: first.start_col,
            attached_code,
            in_function: false,
            function_name: String::new(),
            is_directive: false,
        });
    }

    SourceFile {
        path: path.to_string(),
        lang: spec.name,
        lines,
        comments,
        code_line_count: code_rows.iter().filter(|&&c| c).count(),
        code_lines,
        comment_line_count: comment_rows.iter().filter(|&&c| c).count(),
    }
}

pub fn extract_path(path: &std::path::Path) -> Option<SourceFile> {
    let (spec, _) = spec_for_path(path.to_str()?)?;
    let source = std::fs::read_to_string(path).ok()?;
    Some(extract_source(&path.to_string_lossy(), &source, spec))
}
